File: backend/app/schemas/ai.py
from typing import Literal

from pydantic import BaseModel, Field


ChatFlowType = Literal["station_info", "weather", "plan_route"] | None

ChatActionType = Literal[
    "open_planning",
    "open_route_text",
    "open_route_map",
    "open_stations",
    "open_station_detail",
    "open_help",
    "open_ticket_guide",
    "open_concession_guide",
    "open_privacy",
    "open_preference",
    "compute_route",
]

ChatBlockType = Literal[
    "heading",
    "paragraph",
    "bullets",
    "numbered",
    "key_values",
    "callout",
    "sources",
    "hours",
    "place_cards",
]

ResponseSourceType = Literal["flow", "db", "api", "gemini", "gemini_maps"]

CalloutTone = Literal["info", "warning", "success"]
KeyValueEmphasis = Literal["supported", "not_supported", "unknown", "neutral", "highlight", "route_endpoint"]


class KeyValueRow(BaseModel):
    label: str
    value: str
    emphasis: KeyValueEmphasis | None = None


class SourceLink(BaseModel):
    title: str
    url: str
    org: str | None = None


class ChatBlock(BaseModel):
    type: ChatBlockType
    text: str | None = None
    items: list[str] | None = None
    rows: list[KeyValueRow] | None = None
    tone: CalloutTone | None = None
    links: list[SourceLink] | None = None


class ChatAction(BaseModel):
    type: ChatActionType
    station_id: str | None = None
    station_name: str | None = None
    origin_name: str | None = None
    destination_name: str | None = None
    station_search_query: str | None = None
    departure_time: str | None = None
    origin_lat: float | None = None
    origin_lon: float | None = None
    origin_google_place_id: str | None = None
    destination_lat: float | None = None
    destination_lon: float | None = None
    destination_google_place_id: str | None = None


class AIConversationResponse(BaseModel):
    conversation_id: str


UiLanguageCode = Literal["EN", "BM"]


class AIMessageRequest(BaseModel):
    message: str = Field(min_length=1)
    ui_language: UiLanguageCode = "EN"
    has_current_route: bool = False
    origin_name: str | None = None
    destination_name: str | None = None
    selected_station_id: str | None = None
    selected_station_name: str | None = None
    chat_flow: ChatFlowType = None
    flow_slots: dict[str, str] = Field(default_factory=dict)


class AIMessageResponse(BaseModel):
    conversation_id: str
    answer: str
    answer_blocks: list[ChatBlock] = Field(default_factory=list)
    in_scope: bool
    actions: list[ChatAction] = Field(default_factory=list)
    chat_flow: ChatFlowType = None
    flow_slots: dict[str, str] = Field(default_factory=dict)
    response_source: ResponseSourceType | None = None

File: frontend/src/types/ai.ts
export type ChatFlowType = 'station_info' | 'weather' | 'plan_route' | null;

export type ChatActionType =
  | 'open_planning'
  | 'open_route_text'
  | 'open_route_map'
  | 'open_stations'
  | 'open_station_detail'
  | 'open_help'
  | 'open_ticket_guide'
  | 'open_concession_guide'
  | 'open_privacy'
  | 'open_preference'
  | 'compute_route';

export type ChatBlockType =
  | 'heading'
  | 'paragraph'
  | 'bullets'
  | 'numbered'
  | 'key_values'
  | 'callout'
  | 'sources'
  | 'hours'
  | 'place_cards';

export type ResponseSourceType = 'flow' | 'db' | 'api' | 'gemini' | 'gemini_maps';

export type CalloutTone = 'info' | 'warning' | 'success';
export type KeyValueEmphasis =
  | 'supported'
  | 'not_supported'
  | 'unknown'
  | 'neutral'
  | 'highlight'
  | 'route_endpoint';

export interface KeyValueRow {
  label: string;
  value: string;
  emphasis?: KeyValueEmphasis | null;
}

export interface SourceLink {
  title: string;
  url: string;
  org?: string | null;
}

export interface ChatBlock {
  type: ChatBlockType;
  text?: string | null;
  items?: string[] | null;
  rows?: KeyValueRow[] | null;
  tone?: CalloutTone | null;
  links?: SourceLink[] | null;
}

export interface ChatAction {
  type: ChatActionType;
  station_id?: string | null;
  station_name?: string | null;
  origin_name?: string | null;
  destination_name?: string | null;
  station_search_query?: string | null;
  departure_time?: string | null;
  origin_lat?: number | null;
  origin_lon?: number | null;
  origin_google_place_id?: string | null;
  destination_lat?: number | null;
  destination_lon?: number | null;
  destination_google_place_id?: string | null;
}

export interface AIConversationResponse {
  conversation_id: string;
}

export interface AIMessageContext {
  has_current_route?: boolean;
  origin_name?: string | null;
  destination_name?: string | null;
  selected_station_id?: string | null;
  selected_station_name?: string | null;
  chat_flow?: ChatFlowType;
  flow_slots?: Record<string, string>;
}

export interface AIMessageResponse {
  conversation_id: string;
  answer: string;
  answer_blocks: ChatBlock[];
  in_scope: boolean;
  actions: ChatAction[];
  chat_flow?: ChatFlowType;
  flow_slots?: Record<string, string>;
  response_source?: ResponseSourceType | null;
}
